source/Weather.cpp

```diff
@@ -85,12 +85,15 @@ void Weather::Step(vector<Visual> &visuals)
 	totalAmount *= currentStrength;
 	visuals.reserve(visuals.size() + totalAmount);
 
+	// If a hazard is system-wide then use the center of the screen as the
+	// origin for new effects. Damage is still dealt from the weather's origin.
```

Comment:
This doesn't feel right; the default origin of effects should be the origin of the hazard.
If we also want hazard-specific effects that act on the center of the screen, then that could be an additional "screen center effect" keyword on the hazard. (I can imagine that that might work nicely for hazards like fog.)
If we also want hazard-specific effects that act on the center of the flagship, then that could be an additional "flagship effect" keyword on the hazard. (This is in a practical sense different from the "screen center effect" when the flagship gets destroyed, the "flagship effect" would disappear, but the "screen center effect" would stay.)

As a content-creator I want to be able to create different effects for the following system-wide hazards combinations:
- origin in the system center, and an effect that is only in the system center
- origin in the system center, and an effect that is only around the flagship
- origin in the system center, and an effect that is only around the screen center
- origin around the flagship, and an effect that is only around the flagship
- origin around the flagship, and an effect that is only around the screen center

Reply:
> This doesn't feel right; the default origin of effects should be the origin of the hazard.

Yeah that's true. This is effectively a workaround for the fact that we can't generate infinite effects. Effects are only generated around the flagship so that the hazard is basically infinite, without actually generating infinite effects.

> (This is in a practical sense different from the "screen center effect" when the flagship gets destroyed, the "flagship effect" would disappear, but the "screen center effect" would stay.)

Hazards are "environmental" and as such they shouldn't simply stop when the flagship gets destroyed. But also a flagship getting destroyed is not currently something that players will experience for more than 10 seconds, since it's game over at that point. So I don't really see the need to distinguish them.

> - origin around the flagship, and an effect that is only around the flagship
> - origin around the flagship, and an effect that is only around the screen center

Those two don't really make sense. Hazards that are emitted from the flagship? This is especially problematic since hazards simply describe the hazard and its effects without specifying an origin. So how would this be defined inside the data files, since it can't be a child to a hazard node.

Reply:
> This doesn't feel right; the default origin of effects should be the origin of the hazard.

The default origin for effects *is* the origin of the hazard. It is only when the hazard is system-wide that the effects get generated around the flagship.

> If we also want hazard-specific effects that act on the center of the screen, then that could be an additional "screen center effect" keyword on the hazard. (I can imagine that that might work nicely for hazards like fog.)

Why should that be divorced from the `"system-wide"` tag? Environmental effects should be used to denote where a weather event is having an impact. This is currently the case where the region that environmental effects get generated is the same region where damage is dealt by hazards. When a hazard is made `"system-wide"`, this damage region then extends to the entire system regardless of where the player goes, but since we can't create effects everywhere, `"system-wide"` also moves the creation of environmental effects to around the screen center (i.e. around the player).

> origin around the flagship, and an effect that is only around the flagship
> origin around the flagship, and an effect that is only around the screen center

What would the explanation be for an environmental hazard that *only* originates from the player's flagship and not any other ship in the system? Why would the player's flagship ever get singled out in such a way by environmental hazards?

Reply:
> What would the explanation be for an environmental hazard that _only_ originates from the player's flagship and not any other ship in the system? Why would the player's flagship ever get singled out in such a way by environmental hazards?

Just for some context, not for this PR; I can think of some stories where this might make sense:
- The player took some drugs or some alcoholic drinks from his/her cargohold and now the ships around the flagship suddenly seem to be getting attracted and pushed away from the flagship periodically. :)
- Or the player entered some system first and some effect (magnetic particles) attach to the first ship that entered the system.

Reply:
It's worth considering how one might support "directional effects" that convey the orientation to the hazard's center when the hazard is marked system-wide.

For example, waves / "space ripples" that emanate from a particular point (perhaps where damage is strongest or weakest).

I believe this is still possible for system-wide and localized effects; it just requires a weather keyword to inform the choice of preliminary visual velocity (currently always 0) and facing (currently always random).